cpp/src/community/louvain.cuh

```diff
@@ -0,0 +1,639 @@
+
+      best_modularity = new_Q;
+
+      shrink_graph(current_graph, d_cluster_vec);
+
+      num_level++;
+    }
+
+    timer_display(std::cout);
+
+    return std::make_pair(num_level, best_modularity);
+  }
+
+ protected:
+  void timer_start(std::string const &region)
+  {
+#ifdef TIMING
+    hr_timer_.start(region);
+#endif
+  }
+
+  void timer_stop(cudaStream_t stream)
+  {
+#ifdef TIMING
+    CUDA_TRY(cudaStreamSynchronize(stream));
+    hr_timer_.stop();
+#endif
+  }
+
+  void timer_display(std::ostream &os)
+  {
+#ifdef TIMING
+    hr_timer_.display(os);
+#endif
+  }
+
+ public:
+  void compute_vertex_and_cluster_weights(graph_type const &graph)
+  {
+    timer_start("compute_vertex_and_cluster_weights");
+
+    edge_t const *d_offsets     = graph.offsets;
+    vertex_t const *d_indices   = graph.indices;
+    weight_t const *d_weights   = graph.edge_data;
+    weight_t *d_vertex_weights  = vertex_weights_v_.data().get();
+    weight_t *d_cluster_weights = cluster_weights_v_.data().get();
+
+    //
+    // MNMG:  copy_v_transform_reduce_out_nbr, then copy
+    //
+    thrust::for_each(
+      rmm::exec_policy(stream_)->on(stream_),
+      thrust::make_counting_iterator<edge_t>(0),
+      thrust::make_counting_iterator<edge_t>(graph.number_of_vertices),
+      [d_offsets, d_indices, d_weights, d_vertex_weights, d_cluster_weights] __device__(
+        vertex_t src) {
+        weight_t sum{0.0};
+
+        for (edge_t i = d_offsets[src]; i < d_offsets[src + 1]; ++i) { sum += d_weights[i]; }
+
+        d_vertex_weights[src]  = sum;
+        d_cluster_weights[src] = sum;
+      });
+
+    timer_stop(stream_);
+  }
+
+  virtual weight_t update_clustering(weight_t total_edge_weight,
+                                     weight_t resolution,
+                                     graph_type const &graph)
+  {
+    timer_start("update_clustering");
+
+    //
+    // MNMG: This is the hard one, see writeup
+    //
+    rmm::device_vector<vertex_t> next_cluster_v(cluster_v_);
+    rmm::device_vector<weight_t> delta_Q_v(graph.number_of_edges);
+    rmm::device_vector<vertex_t> cluster_hash_v(graph.number_of_edges);
+    rmm::device_vector<weight_t> old_cluster_sum_v(graph.number_of_vertices);
+
+    vertex_t *d_cluster_hash         = cluster_hash_v.data().get();
+    vertex_t *d_cluster              = cluster_v_.data().get();
+    weight_t const *d_vertex_weights = vertex_weights_v_.data().get();
+    weight_t *d_cluster_weights      = cluster_weights_v_.data().get();
+    weight_t *d_delta_Q              = delta_Q_v.data().get();
+
+    weight_t new_Q = modularity(total_edge_weight, resolution, graph, cluster_v_.data().get());
+
+    weight_t cur_Q = new_Q - 1;
+
+    // To avoid the potential of having two vertices swap clusters
+    // we will only allow vertices to move up (true) or down (false)
+    // during each iteration of the loop
+    bool up_down = true;
+
+    while (new_Q > (cur_Q + 0.0001)) {
```

Comment:
Please consider using a `for_each()` or `transform()` on the sequence `{0,...,max_level-1}` rather than a while loop.

Reply:
This loop runs an indeterminate number of times (nothing like a max_level).

Reply:
Please ignore then.